File: ios/interfaces/analyst/SNetworkAnalyst.m
//
//  SNetworkAnalyst.m
//  Supermap
//
//  Created by Shanglong Yang on 2019/6/24.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "SNetworkAnalyst.h"

@implementation History : NSObject

-(id)init {
    self = [super init];
    _history = [[NSMutableArray alloc] init];
    return self;
}
-(int)getCount {
    return _history.count;
}
-(NSArray *)getHistory {
    return [_history subarrayWithRange:NSMakeRange(0, _currentIndex)];
}
-(NSDictionary *)get:(int)index {
    return _history[index];
}
-(void)addHistory:(NSDictionary *)obj {
    if (_currentIndex < (NSInteger)_history.count - 1) {
        NSArray* temp = [_history subarrayWithRange:NSMakeRange(0, _currentIndex + 1)];
        _history = [[NSMutableArray alloc] initWithArray:temp];
    }
    [_history addObject:obj];
    _currentIndex = (NSInteger)_history.count - 1;
}
-(void)removeObjectAtIndex:(int)index {
    [_history removeObjectAtIndex:index];
    _currentIndex--;
}
-(void)removeObject:(NSObject *)obj {
    [_history removeObject:obj];
    _currentIndex--;
}
-(void)clear {
    [_history removeAllObjects];
    _currentIndex = -1;
}
-(int)undo {
    int preIndex = _currentIndex >= 0 ? _currentIndex : -1;
    if (_currentIndex > -1) {
        _currentIndex--;
    }
    return preIndex;
}
-(int)redo {
    int historyCount = (NSInteger)_history.count;
    int preIndex = _currentIndex < historyCount - 1 ? _currentIndex : (historyCount - 1);
    if (_currentIndex < historyCount - 1) {
        _currentIndex++;
    }
    return preIndex;
}
@end

@implementation SNetworkAnalyst
RCT_EXPORT_MODULE();

- (GeoStyle *)getGeoStyle:(Size2D *)size2D color:(Color *)color {
    GeoStyle* style = [[GeoStyle alloc] init];
    [style setMarkerSize:size2D];
    [style setLineColor:color];
    [style setLineWidth:1];
    return style;
}

- (void)displayResult:(NSArray *)ids selection:(Selection *)selection {
    if (selection) {
        TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
        for (int i = 0; i < ids.count; i++) {
            [selection add:[(NSNumber *)ids[i] intValue]];
        }
        Recordset* recordset = [selection toRecordset];
        [recordset moveFirst];
        while (![recordset isEOF]) {
            Geometry* geometry = recordset.geometry;
            GeoStyle* style = [self getGeoStyle:[[Size2D alloc] initWithWidth:10 Height:10] color:[[Color alloc] initWithR:255 G:105 B:0]];
            [geometry setStyle:style];
            [trackingLayer addGeometry:geometry WithTag:@""];
            [recordset moveNext];
        }
        [recordset dispose];
        [[SMap singletonInstance].smMapWC.mapControl.map refresh];
    }
}

- (void)clear {
    if (selection) {
        [selection clear];
    }
    [[SMap singletonInstance].smMapWC.mapControl.map.trackingLayer clear];
}

- (void)clearRoutes:(Selection *)selection {
    if (selection) {
        [selection clear];
    }
    TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
    int routeIndex = [trackingLayer indexof:@"route"];
    while (routeIndex >= 0) {
        [trackingLayer removeAt:routeIndex];
        routeIndex = [trackingLayer indexof:@"route"];
    }
}

- (int)selectNode:(NSDictionary *)point layer:(Layer *)nodeLayer geoStyle:(GeoStyle *)geoStyle tag:(NSString *)tag {
    int ID = -1;
//    if (!elementIDs) {
//        elementIDs = [[NSMutableArray alloc] init];
//    }
    double x = [(NSNumber *)[point objectForKey:@"x"] doubleValue];
    double y = [(NSNumber *)[point objectForKey:@"y"] doubleValue];
    CGPoint p = CGPointMake(x, y);
    Selection* hitSelection = [nodeLayer hitTestEx:p With:20];
    
    if (hitSelection && hitSelection.getCount > 0) {
        Recordset* rs = hitSelection.toRecordset;
        GeoPoint* gPoint = (GeoPoint *)rs.geometry;
        Point2D* p2D = [[Point2D alloc] initWithX:gPoint.getX Y:gPoint.getY];
        ID = (int)rs.ID;
//        [elementIDs addObject:@(ID)];

        BOOL isExsit = [self pointIsExist:p2D];
        if (isExsit) return -1;
        
        GeoStyle* style = geoStyle;
        if (!style) {
            style = [[GeoStyle alloc] init];
            [style setMarkerSize:[[Size2D alloc] initWithWidth:10 Height:10]];
            [style setLineColor:[[Color alloc] initWithR:255 G:105 B:0]];
            [style setMarkerSymbolID:3614];
        }
        [gPoint setStyle:style];
        
        TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
        [trackingLayer addGeometry:gPoint WithTag:tag];
        
//        NSMutableDictionary* nodeData = [[NSMutableDictionary alloc] init];
//        [nodeData setObject:tag forKey:@"tag"];
//        [nodeData setObject:point forKey:@"node"];
//        if (history == nil) {
//            history = [[History alloc] init];
//        }
//        [history addHistory:nodeData];
        
        [gPoint dispose];
        [rs close];
        [rs dispose];
    }
    return ID;
}

- (NSDictionary *)selectNodeWithPoint:(NSDictionary *)point layer:(Layer *)nodeLayer geoStyle:(GeoStyle *)geoStyle tag:(NSString *)tag {
    int ID = -1;
    Point2D* p2D = nil;
    double x = [(NSNumber *)[point objectForKey:@"x"] doubleValue];
    double y = [(NSNumber *)[point objectForKey:@"y"] doubleValue];
    CGPoint p = CGPointMake(x, y);
    Selection* hitSelection = [nodeLayer hitTestEx:p With:20];
    
    if (hitSelection && hitSelection.getCount > 0) {
        Recordset* rs = hitSelection.toRecordset;
        GeoPoint* gPoint = (GeoPoint *)rs.geometry;
        ID = (int)rs.ID;
        p2D = [[Point2D alloc] initWithX:gPoint.getX Y:gPoint.getY];
        
        BOOL isExsit = [self pointIsExist:p2D];
        if (isExsit) return nil;
        
        GeoStyle* style = geoStyle;
        if (!style) {
            style = [[GeoStyle alloc] init];
            [style setMarkerSize:[[Size2D alloc] initWithWidth:10 Height:10]];
            [style setLineColor:[[Color alloc] initWithR:255 G:105 B:0]];
            [style setMarkerSymbolID:3614];
        }
        [gPoint setStyle:style];
        
        TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
        [trackingLayer addGeometry:gPoint WithTag:tag];
        
//        NSMutableDictionary* nodeData = [[NSMutableDictionary alloc] init];
//        [nodeData setObject:tag forKey:@"tag"];
//        [nodeData setObject:point forKey:@"node"];
//        if (history == nil) {
//            history = [[History alloc] init];
//        }
//        [history addHistory:nodeData];
        
        [gPoint dispose];
        [rs close];
        [rs dispose];
    }
    return @{
        @"ID": @(ID),
        @"point": p2D,
    };
}

- (Point2D *)selectPoint:(NSDictionary *)point layer:(Layer *)nodeLayer geoStyle:(GeoStyle *)geoStyle tag:(NSString *)tag {
//    int ID = -1;
    double x = [(NSNumber *)[point objectForKey:@"x"] doubleValue];
    double y = [(NSNumber *)[point objectForKey:@"y"] doubleValue];
    CGPoint p = CGPointMake(x, y);
    Selection* hitSelection = [nodeLayer hitTestEx:p With:20];
    NSMutableDictionary* pDic = nil;
    Point2D* p2D = nil;
    
    if (hitSelection && hitSelection.getCount > 0) {
        pDic = [[NSMutableDictionary alloc] init];
        Recordset* rs = hitSelection.toRecordset;
        GeoPoint* gPoint = (GeoPoint *)rs.geometry;
        
        p2D = [[Point2D alloc] initWithX:gPoint.getX Y:gPoint.getY];
        
        BOOL isExsit = [self pointIsExist:p2D];
        if (isExsit) return nil;

        GeoStyle* style = geoStyle;
        if (!style) {
            style = [[GeoStyle alloc] init];
            [style setMarkerSize:[[Size2D alloc] initWithWidth:10 Height:10]];
            [style setLineColor:[[Color alloc] initWithR:255 G:105 B:0]];
            [style setMarkerSymbolID:3614];
        }
        [gPoint setStyle:style];
        
        TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
        [trackingLayer addGeometry:gPoint WithTag:tag];
        
        [gPoint dispose];
        [rs close];
        [rs dispose];
    }
    return p2D;
}

- (int)setText:(NSString *)text point:(Point2D *)point textStyle:(TextStyle *)textStyle tag:(NSString *)tag {
    NSString* mText = [NSString stringWithFormat:@" %@", text];
    TextPart* textPart = [[TextPart alloc] initWithTextString:mText anchorPoint:point];
    GeoText* geoText = [[GeoText alloc] initWithTextPart:textPart];
    if (textStyle == nil) {
        textStyle = [[TextStyle alloc]init];
        [textStyle setOutline:YES];
        [textStyle setFontWidth:6];
        [textStyle setFontHeight:8];
        [textStyle setForeColor:[[Color alloc]initWithR:0 G:0 B:0]];
    }
    [textStyle setOutline:YES];
    [textStyle setBackColor:[[Color alloc] initWithR:255 G:255 B:255]];
    [textStyle setBold:YES];
    [geoText setTextStyle:textStyle];
    
    TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
    int index = [trackingLayer addGeometry:geoText WithTag:tag];
    [[SMap singletonInstance].smMapWC.mapControl.map refresh];
    return index;
}

//- (Point2D *)selectByPoint:(NSDictionary *)point layer:(Layer *)nodeLayer geoStyle:(GeoStyle *)geoStyle tag:(NSString *)tag {
//    Point2D* point2D = nil;
//    //    if (!elementIDs) {
//    //        elementIDs = [[NSMutableArray alloc] init];
//    //    }
//    double x = [(NSNumber *)[point objectForKey:@"x"] doubleValue];
//    double y = [(NSNumber *)[point objectForKey:@"y"] doubleValue];
//    CGPoint p = CGPointMake(x, y);
//    Selection* hitSelection = [nodeLayer hitTestEx:p With:20];
//    
//    if (hitSelection && hitSelection.getCount > 0) {
//        Recordset* rs = hitSelection.toRecordset;
//        GeoPoint* gPoint = (GeoPoint *)rs.geometry;
//        point2D = gPoint.getInnerPoint;
//        //        [elementIDs addObject:@(ID)];
//        
//        GeoStyle* style = geoStyle;
//        if (!style) {
//            style = [[GeoStyle alloc] init];
//            [style setMarkerSize:[[Size2D alloc] initWithWidth:10 Height:10]];
//            [style setLineColor:[[Color alloc] initWithR:255 G:105 B:0]];
//            [style setMarkerSymbolID:3614];
//        }
//        [gPoint setStyle:style];
//        
//        TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
//        [trackingLayer addGeometry:gPoint WithTag:tag];
//        [[SMap singletonInstance].smMapWC.mapControl.map refresh];
//        
//        [gPoint dispose];
//        [rs close];
//        [rs dispose];
//    }
//    return point2D;
//}

- (void)removeTagFromTrackingLayer:(NSString *)tag {
    TrackingLayer* trackingLayer = [SMap singletonInstance].smMapWC.mapControl.map.trackingLayer;
    for (int i = 0; i < trackingLayer.count; i++) {
        NSString* currentTag = [trackingLayer getTagAt:i];
        if ([currentTag isEqualToString:tag]) {
            [trackingLayer removeAt:i];
            break;
        }
    }
}

- (BOOL)pointIsExist:(Point2D *)point {
    if (!points) points = [[Point2Ds alloc] init];
    if (!barrierPoints) barrierPoints = [[Point2Ds alloc] init];
    for (int i = 0; i < barrierPoints.getCount; i++) {
        Point2D* pt = [barrierPoints getItem:i];
        if (pt.x == point.x && pt.y == point.y) {
            return YES;
        }
    }
    for (int i = 0; i < points.getCount; i++) {
        Point2D* pt = [points getItem:i];
        if (pt.x == point.x && pt.y == point.y) {
            return YES;
        }
    }
    if (startPoint && startPoint.x == point.x && startPoint.y == point.y) {
        return YES;
    }
    if (endPoint && endPoint.x == point.x && endPoint.y == point.y) {
        return YES;
    }
    return NO;
}
@end

File: android/src/main/java/com/supermap/interfaces/analyst/SNetworkAnalyst.java
package com.supermap.interfaces.analyst;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReadableMap;
import com.supermap.data.Color;
import com.supermap.data.GeoPoint;
import com.supermap.data.GeoStyle;
import com.supermap.data.GeoText;
import com.supermap.data.Geometry;
import com.supermap.data.Point;
import com.supermap.data.Point2D;
import com.supermap.data.Point2Ds;
import com.supermap.data.Recordset;
import com.supermap.data.Size2D;
import com.supermap.data.TextPart;
import com.supermap.data.TextStyle;
import com.supermap.interfaces.mapping.SMap;
import com.supermap.mapping.Layer;
import com.supermap.mapping.Selection;
import com.supermap.mapping.TrackingLayer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import de.javagl.obj.Obj;

class History {
    private int currentIndex;

    private ArrayList<HashMap> history;

    public History() {
        history = new ArrayList<>();
    }

    public int getCount() {
        return history.size();
    }

    public int getCurrentIndex() {
        return currentIndex;
    }

    public ArrayList<HashMap> getAllHistory() {
        return history;
    }

    public ArrayList<HashMap> getHistory() {
        return (ArrayList<HashMap>)history.subList(0, currentIndex);
    }

    public HashMap get(int index) {
        return history.get(index);
    }

    public void addHistory(HashMap obj) {
        if (currentIndex < history.size() - 1) {
            List list = history.subList(0, currentIndex + 1);
            ArrayList<HashMap> temp = new ArrayList<>();
            for (int i = 0; i < list.size(); i++) {
                temp.add((HashMap) list.get(i));
            }
            history = temp;
        }
        history.add(obj);
        currentIndex = history.size() - 1;
    }

    public void remove(int index) {
        history.remove(index);
        currentIndex -= 1;
    }

    public void remove(Object obj) {
        history.remove(obj);
        currentIndex -= 1;
    }

    public void clear() {
        history.clear();
        currentIndex = -1;
    }

    public int undo() {
        int preIndex = currentIndex >= 0 ? currentIndex : -1;
        if (currentIndex > -1) {
            currentIndex--;
        }
        return preIndex;
    }

    public int redo() {
        int preIndex = currentIndex < history.size() - 1 ? currentIndex : (history.size() - 1);
        if (currentIndex < history.size() - 1) {
            currentIndex++;
        }
        return preIndex;
    }
}

public class SNetworkAnalyst extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "SNetworkAnalyst";
    public Layer layer;
    public Layer nodeLayer;
    public Selection selection;
    public ArrayList<Integer> elementIDs;
    public int startNodeID;
    public int endNodeID;
    public Point2D startPoint;
    public Point2D endPoint;
    public ArrayList<Integer> nodes = null;
    public ArrayList<Integer> barrierNodes = null;
    public Point2Ds points = null;
    public Point2Ds barrierPoints = null;
    public ArrayList<Integer> middleNodeIDs;
    public History history;
    ReactContext mReactContext;

    public static String routeTag = "route";

    public SNetworkAnalyst(ReactApplicationContext context) {
        super(context);
        mReactContext = context;
        history = new History();
    }

    public void close() {
        layer = null;
        nodeLayer = null;
        selection.clear();
        selection = null;
        elementIDs.clear();
        elementIDs = null;
        startPoint = null;
        endPoint = null;
        middleNodeIDs.clear();
        middleNodeIDs = null;
        history.clear();
        history = null;
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    private GeoStyle getGeoStyle(Size2D size2D, Color color) {
        GeoStyle geoStyle = new GeoStyle();
        geoStyle.setMarkerSize(size2D);
        geoStyle.setLineColor(color);
        return geoStyle;
    }

    public void displayResult(int[] ids, Selection selection) {
        if (selection != null) {
            TrackingLayer trackingLayer = SMap.getInstance().getSmMapWC().getMapControl().getMap().getTrackingLayer();
            for (int i = 0; i < ids.length; i++) {
                selection.add(ids[i]);
            }
            Recordset recordset = selection.toRecordset();
            while (!recordset.isEOF()) {
                Geometry geometry = recordset.getGeometry();
                GeoStyle style = getGeoStyle(new Size2D(10, 10), new Color(255, 105, 0));
                geometry.setStyle(style);
                trackingLayer.add(geometry, "");
                recordset.moveNext();
            }
            SMap.getInstance().getSmMapWC().getMapControl().getMap().refresh();
        }
    }

    public void clear() {
        if (this.selection != null) {
            this.selection.clear();
        }
        SMap.getInstance().getSmMapWC().getMapControl().getMap().getTrackingLayer().clear();
    }

    public void clearRoutes() {
        if (this.selection != null) {
            this.selection.clear();
        }
        TrackingLayer trackingLayer = SMap.getInstance().getSmMapWC().getMapControl().getMap().getTrackingLayer();
        int routeIndex = trackingLayer.indexOf(routeTag);
        while (routeIndex >= 0) {
            trackingLayer.remove(routeIndex);
            routeIndex = trackingLayer.indexOf(routeTag);
        }
    }

    public int selectNode(ReadableMap point, Layer nodeLayer, GeoStyle geoStyle, String tag) {
        int ID = -1;

        double x = point.getDouble("x");
        double y = point.getDouble("y");
        Point p = new Point((int)x, (int)y);

        Selection hitSelection = nodeLayer.hitTestEx(p, 20);
        if (hitSelection != null && hitSelection.getCount() > 0) {
            Recordset rs = hitSelection.toRecordset();
            GeoPoint geoPoint = (GeoPoint)rs.getGeometry();
            Point2D point2D = new Point2D(geoPoint.getX(), geoPoint.getY());
            ID = rs.getID();
            // 已存在，返回null
            Boolean isExist = pointIsExist(point2D);
            if (isExist) return -1;

            GeoStyle style = geoStyle;
            if (style == null) {
                style = getGeoStyle(new Size2D(10, 10), new Color(255, 105, 0));
                style.setMarkerSymbolID(3614);
            }
            geoPoint.setStyle(style);

            TrackingLayer trackingLayer = SMap.getInstance().getSmMapWC().getMapControl().getMap().getTrackingLayer();
            trackingLayer.add(geoPoint, tag);

            HashMap nodeMap = new HashMap();
            HashMap node = new HashMap();
            node.put("x", x);
            node.put("y", y);
            nodeMap.put("tag", tag);
            nodeMap.put("node", node);
            if (history == null) {
                history = new History();
            }
            history.addHistory(nodeMap);

            geoPoint.dispose();
            rs.close();
            rs.dispose();
        }
        return ID;
    }

    public Point2D selectPoint(ReadableMap point, Layer nodeLayer, GeoStyle geoStyle, String tag) {
        Point2D point2D = null;
        int ID = -1;

        double x = point.getDouble("x");
        double y = point.getDouble("y");
        Point p = new Point((int)x, (int)y);

        Selection hitSelection = nodeLayer.hitTestEx(p, 20);
        if (hitSelection != null && hitSelection.getCount() > 0) {
            Recordset rs = hitSelection.toRecordset();
            GeoPoint geoPoint = (GeoPoint)rs.getGeometry();
//            point2D = geoPoint.getInnerPoint();
//            Point p2 = SMap.getInstance().getSmMapWC().getMapControl().getMap().mapToPixel(point2D);
//            map.putInt("x", p2.getX());
//            map.putInt("x", p2.getY());
            point2D = new Point2D(geoPoint.getX(), geoPoint.getY());

            // 已存在，返回null
            Boolean isExist = pointIsExist(point2D);
            if (isExist) return null;

            GeoStyle style = geoStyle;
            if (style == null) {
                style = getGeoStyle(new Size2D(10, 10), new Color(255, 105, 0));
                style.setMarkerSymbolID(3614);
            }
            geoPoint.setStyle(style);

            TrackingLayer trackingLayer = SMap.getInstance().getSmMapWC().getMapControl().getMap().getTrackingLayer();
            trackingLayer.add(geoPoint, tag);

            geoPoint.dispose();
            rs.close();
            rs.dispose();
        }
        return point2D;
    }

    public int setText(String text, Point2D point, TextStyle textStyle, String tag) {
        TextPart textPart = new TextPart(" " + text, point);
        GeoText geoText = new GeoText(textPart);
        if (textStyle == null) {
            textStyle = new TextStyle();
            textStyle.setOutline(true);
            textStyle.setFontWidth(6);
            textStyle.setFontHeight(8);
            textStyle.setForeColor(new Color(0, 0, 0));
        }
        textStyle.setOutline(true);
        textStyle.setBackColor(new Color(255, 255, 255));
        textStyle.setBold(true);
        geoText.setTextStyle(textStyle);

        TrackingLayer trackingLayer = SMap.getInstance().getSmMapWC().getMapControl().getMap().getTrackingLayer();
        int index = trackingLayer.add(geoText, tag);
        return index;
    }

    public void removeTagFromTrackingLayer(String tag) {
        TrackingLayer trackingLayer = SMap.getInstance().getSmMapWC().getMapControl().getMap().getTrackingLayer();
        for (int i = 0; i < trackingLayer.getCount(); i++) {
            String currentTag = trackingLayer.getTag(i);
            if (currentTag.equals(tag)) {
                trackingLayer.remove(i);
                break;
            }
        }
    }

    public boolean pointIsExist(Point2D point) {
        if (points == null) points = new Point2Ds();
        if (barrierPoints == null) barrierPoints = new Point2Ds();
        for (int i = 0; i < barrierPoints.getCount(); i++) {
            Point2D pt = barrierPoints.getItem(i);
            if (pt.getX() == point.getX() && pt.getY() == point.getY()) {
                return true;
            }
        }
        for (int i = 0; i < points.getCount(); i++) {
            Point2D pt = points.getItem(i);
            if (pt.getX() == point.getX() && pt.getY() == point.getY()) {
                return true;
            }
        }
        if (startPoint != null && startPoint.getX() == point.getX() && startPoint.getY() == point.getY()) {
            return true;
        }
        if (endPoint != null && endPoint.getX() == point.getX() && endPoint.getY() == point.getY()) {
            return true;
        }
        return false;
    }

//    public WritableMap undo() {
//        int preIndex = history.undo();
//        WritableMap node = history.get(preIndex);
//        return node;
//    }
//
//    public void redo() {
//        history.redo();
//    }
}